### PR TITLE
Adds new oncall service feedback table

### PR DESCRIPTION
**Pull request by whitdog47**
This PR adds a new oncall end-of-shift feedback table, filterable by project. An additional change allows the post-shift hours to be fractional.

Screenshot 2023-10-06 at 11 43 03 AM

### CODE REVIEW

File: src/dispatch/static/dispatch/src/service_feedback/Table.vue [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            :page.sync="page"
+            :items-per-page.sync="itemsPerPage"
+            :sort-by.sync="sortBy"
+            :sort-desc.sync="descending"
+            :loading="loading"
+            loading-text="Loading... Please wait"
+          >
+            <template #item.individual="{ item }">
+              <participant
+                v-if="item.individual"
```
Comment (metroid-samus) [inline src/dispatch/static/dispatch/src/service_feedback/Table.vue:38]: We should populate this item in this column as "Anonymous" if the participant submitted their feedback anonymously.

Comment (whitdog47) [reply src/dispatch/static/dispatch/src/service_feedback/Table.vue:38]: Good idea. Added!

File: src/dispatch/feedback/service/views.py [hunk 2]
```diff
@@ -12,3 +19,31 @@
...
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail=[{"msg": "A feedback entry with this id does not exist."}],
+        )
+    return feedback
+
+
+@router.delete(
+    "/{service_feedback_id}",
+    response_model=None,
+    dependencies=[Depends(PermissionsDependency([OrganizationAdminPermission]))],
```
Comment (mvilanova) [inline src/dispatch/feedback/service/views.py:39]: Should we allow the creator of the feedback to delete it as well, unless it's anonymous?

File: src/dispatch/feedback/service/models.py [hunk 3]
```diff
@@ -21,14 +20,17 @@ class ServiceFeedback(TimeStampMixin, FeedbackMixin, Base):
     feedback = Column(String)
     rating = Column(String)
     schedule = Column(String)
-    hours = Column(Integer)
+    hours = Column(Numeric(precision=10, scale=2))
     shift_start_at = Column(DateTime)
     shift_end_at = Column(DateTime)
 
     # Relationships
     individual_contact_id = Column(Integer, ForeignKey("individual_contact.id"))
```
Comment (mvilanova) [inline src/dispatch/feedback/service/models.py:28]: If the individual were to get deleted, do you want to keep the feedback or delete it? If the latter, then you want to add `ondelete="CASCADE"`

`individual_id = Column(Integer, ForeignKey("individual_contact.id", ondelete="CASCADE"))`

File: src/dispatch/feedback/service/models.py [hunk 4]
```diff
@@ -21,14 +20,17 @@ class ServiceFeedback(TimeStampMixin, FeedbackMixin, Base):
...
     individual = relationship("IndividualContact")
```
Comment (mvilanova) [inline src/dispatch/feedback/service/models.py:29]: I'd move this relationship to the `IndividualContact` model and make it backref, like the one we have for events

`events = relationship("Event", backref="individual")`

File: src/dispatch/feedback/service/models.py [hunk 5]
```diff
@@ -37,20 +39,18 @@ class ServiceFeedback(TimeStampMixin, FeedbackMixin, Base):
         )
     )
 
-    @hybrid_property
-    def project(self):
-        return self.individual.project
```
Comment (mvilanova) [inline src/dispatch/feedback/service/models.py:42]: Did this not work and that's why you had to add a project column?

**Review by metroid-samus — APPROVED** [review]

File: src/dispatch/static/dispatch/src/router/config.js [hunk 6]
```diff
@@ -298,9 +298,15 @@ export const protectedRoute = [
         {
           path: "/:organization/feedback",
```
Comment (mvilanova) [inline src/dispatch/static/dispatch/src/router/config.js:299]:
```suggestion
          path: "/:organization/feedback/incident",
```

File: src/dispatch/static/dispatch/src/router/config.js [hunk 7]
```diff
@@ -298,9 +298,15 @@ export const protectedRoute = [
...
           name: "FeedbackTable",
-          meta: { title: "Feedback" },
+          meta: { title: "Incident feedback", group: "feedback" },
           component: () => import("@/feedback/Table.vue"),
         },
+        {
+          path: "/:organization/service-feedback",
```
Comment (mvilanova) [inline src/dispatch/static/dispatch/src/router/config.js:305]:
```suggestion
          path: "/:organization/feedback/service",
```

File: src/dispatch/static/dispatch/src/router/config.js [hunk 8]
```diff
@@ -298,9 +298,15 @@ export const protectedRoute = [
...
+          name: "ServiceFeedbackTable",
+          meta: { title: "Service feedback", group: "feedback" },
+          component: () => import("@/service_feedback/Table.vue"),
```
Comment (mvilanova) [inline src/dispatch/static/dispatch/src/router/config.js:308]: I'd organize the Vue.js code the same we organize the backend code. One directory called `feeback`, which already exists, with two subdirectories called `incident` and `service`. Then here you'd do:

```suggestion
          component: () => import("@/feedback/service/Table.vue"),
```

File: src/dispatch/static/dispatch/src/router/config.js [hunk 9]
```diff
@@ -298,9 +298,15 @@ export const protectedRoute = [
         {
           path: "/:organization/feedback",
           name: "FeedbackTable",
-          meta: { title: "Feedback" },
+          meta: { title: "Incident feedback", group: "feedback" },
           component: () => import("@/feedback/Table.vue"),
```
Comment (mvilanova) [inline src/dispatch/static/dispatch/src/router/config.js:302]: We need to move everything we have in `feedback` to a new subdirectory called `incident`

```suggestion
          component: () => import("@/feedback/incident/Table.vue"),
```

File: src/dispatch/static/dispatch/src/service_feedback/DeleteDialog.vue [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      </v-card-actions>
+    </v-card>
+  </v-dialog>
+</template>
+
+<script>
+import { mapActions } from "vuex"
+import { mapFields } from "vuex-map-fields"
+export default {
+  name: "FeedbackDeleteDialog",
```
Comment (mvilanova) [inline src/dispatch/static/dispatch/src/service_feedback/DeleteDialog.vue:25]:
```suggestion
  name: "ServiceFeedbackDeleteDialog",
```

File: src/dispatch/static/dispatch/src/service_feedback/Table.vue [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+  components: {
+    DeleteDialog,
+    Participant,
+    TableFilterDialog,
+  },
+
+  data() {
+    return {
+      headers: [
+        { text: "Shift End At", value: "shift_end_at", sortable: true },
```
Comment (mvilanova) [inline src/dispatch/static/dispatch/src/service_feedback/Table.vue:98]: I'd format this datetime as well.

File: src/dispatch/static/dispatch/src/store.js [hunk 12]
```diff
@@ -30,6 +30,7 @@ import reference from "@/document/reference/store"
 import runbook from "@/document/runbook/store"
 import search from "@/search/store"
 import service from "@/service/store"
+import service_feedback from "@/service_feedback/store"
```
Comment (mvilanova) [inline src/dispatch/static/dispatch/src/store.js:33]:
```suggestion
import service_feedback from "@/feedback/service/store"
```

File: src/dispatch/feedback/service/models.py (hunk 3 above)
Comment (whitdog47) [reply src/dispatch/feedback/service/models.py:26]: I think we should keep the feedback even if the individual is deleted.

File: src/dispatch/feedback/service/models.py (hunk 5 above)
Comment (whitdog47) [reply src/dispatch/feedback/service/models.py:42]: (discussed offline)

**Review by mvilanova — APPROVED** [review]

**Pull request merged**